test/gtest/uct/test_pending.cc

```diff
@@ -577,3 +578,130 @@ UCS_TEST_SKIP_COND_P(test_uct_pending, pending_fairness,
+                                                       uct);
+
+        return uct_ep_am_short(req->ep, AM_ID, m_pending_hdr, NULL, 0);
+    }
+
+    static void completion_cb(uct_completion_t *self, ucs_status_t c_status) {
+        am_completion_t *comp = ucs_container_of(self, am_completion_t, uct);
+
+        EXPECT_UCS_OK(c_status);
+
+        ucs_status_t status = uct_ep_am_short(comp->ep, AM_ID, m_comp_hdr,
+                                              NULL, 0);
+        EXPECT_TRUE(!UCS_STATUS_IS_ERR(status) ||
+                    (status == UCS_ERR_NO_RESOURCE));
+    }
+
+protected:
+    static const uint64_t m_am_hdr;
+    static const uint64_t m_pending_hdr;
+    static const uint64_t m_comp_hdr;
+    static const uint8_t  AM_ID;
+    static bool m_pend_received;
+};
+
+bool test_uct_pending_ooo::m_pend_received         = false;
+const uint64_t test_uct_pending_ooo::m_am_hdr      = 0x0ul;
+const uint64_t test_uct_pending_ooo::m_pending_hdr = 0x1ul;
+const uint64_t test_uct_pending_ooo::m_comp_hdr    = 0x2ul;
+const uint8_t test_uct_pending_ooo::AM_ID          = 10;
+
+/* Check that pending requests are processed before the sends from
+ * completion callbacks */
+UCS_TEST_SKIP_COND_P(test_uct_pending_ooo, send_ooo_with_comp,
+                     !check_caps(UCT_IFACE_FLAG_AM_SHORT |
+                                 UCT_IFACE_FLAG_AM_ZCOPY |
+                                 UCT_IFACE_FLAG_PENDING))
+{
+    initialize();
+
+    bool comp_received = false;
+    m_pend_received    = false;
+
+    install_handler_sync_or_async(m_e2->iface(), AM_ID,
+                                  am_handler_check_rx_order, &comp_received);
+
+    mapped_buffer sendbuf(32, 0, *m_e1);
+    UCS_TEST_GET_BUFFER_IOV(iov, iovcnt, sendbuf.ptr(), sendbuf.length(),
+                            sendbuf.memh(), 1);
+    am_completion_t comp;
+    comp.uct.func        = completion_cb;
+    comp.uct.count       = 1;
+    comp.ep              = m_e1->ep(0);
+    ucs_status_t status  = uct_ep_am_zcopy(m_e1->ep(0), AM_ID, &m_am_hdr,
+                                           sizeof(m_am_hdr), iov, iovcnt, 0,
+                                           &comp.uct);
+    ASSERT_FALSE(UCS_STATUS_IS_ERR(status));
+
+    ucs_time_t loop_end_limit = ucs_get_time() + ucs_time_from_sec(2);
+    do {
+        status = uct_ep_am_short(m_e1->ep(0), AM_ID, m_am_hdr, NULL, 0);
+    } while (!UCS_STATUS_IS_ERR(status) && (ucs_get_time() < loop_end_limit));
+
+    if (!UCS_STATUS_IS_ERR(status)) {
+        /* got here due to reaching the time limit in the above loop.
+         * couldn't add a request to pending. all sends were successful. */
+        UCS_TEST_MESSAGE << "Can't create out-of-order in the given time.";
+        return;
+    }
+
+    EXPECT_EQ(UCS_ERR_NO_RESOURCE, status);
+    pending_send_request_t *req = pending_alloc_short();
+    status                      = uct_ep_pending_add(m_e1->ep(0), &req->uct, 0);
```

Comment:
in theory pending_add could fail here with BUSY and we need to retry sending

Reply:
This behavior is not supposed to happen, so
- I check that the latest send status is `NO_RESOURCE`
- I check that `uct_ep_pending_add` returns `OK`

Reply:
could ud async thread release some send resources in between?